File: java/Samples/Biometrics/CS/SimpleIrisesSampleCS/IdentifyIris.cs
using System;
using System.Globalization;
using System.IO;
using System.Linq;
using System.Windows.Forms;
using Neurotec.Biometrics;
using Neurotec.Biometrics.Client;
using Neurotec.Images;

namespace Neurotec.Samples
{
	public partial class IdentifyIris : UserControl
	{
		#region Public constructor

		public IdentifyIris()
		{
			InitializeComponent();
		}

		#endregion

		#region Private fields

		private NBiometricClient _biometricClient;
		private NSubject _subject;
		private NSubject[] _subjects;
		private NBiometricStatus _enrollStatus = NBiometricStatus.None;

		private int _defaultFar;

		#endregion

		#region Public properties

		public NBiometricClient BiometricClient
		{
			get { return _biometricClient; }
			set
			{
				_biometricClient = value;
				_defaultFar = _biometricClient.MatchingThreshold;
				cbMatchingFar.Text = Utils.MatchingThresholdToString(_defaultFar);
			}
		}

		#endregion

		#region Private methods

		private void OnEnrollCompleted(IAsyncResult r)
		{
			if (InvokeRequired)
			{
				BeginInvoke(new AsyncCallback(OnEnrollCompleted), r);
			}
			else
			{
				NBiometricTask task = _biometricClient.EndPerformTask(r);
				_enrollStatus = task.Status;
				if (_enrollStatus == NBiometricStatus.Ok)
				{
					btnIdentify.Enabled = IsSubjectValid(_subject);
				}
				else MessageBox.Show(string.Format("Enrollment failed: {0}", _enrollStatus));
			}
		}

		private void OnIdentifyDone(IAsyncResult r)
		{
			if (InvokeRequired)
			{
				BeginInvoke(new AsyncCallback(OnIdentifyDone), r);
			}
			else
			{
				try
				{
					NBiometricStatus status = _biometricClient.EndIdentify(r);
					if (status == NBiometricStatus.Ok || status == NBiometricStatus.MatchNotFound)
					{
						// Matching subjects
						foreach (var result in _subject.MatchingResults)
						{
							listView.Items.Add(new ListViewItem(new[] { result.Id, result.Score.ToString(CultureInfo.InvariantCulture) }));
						}
						foreach (var subject in _subjects)
						{
							bool isMatchingResult = _subject.MatchingResults.Any(result => subject.Id == result.Id);
							if (!isMatchingResult)
								listView.Items.Add(new ListViewItem(new[] { subject.Id, "0" }));
						}
					}
					else MessageBox.Show(string.Format("Identification failed: {0}", status));
				}
				catch (Exception ex)
				{
					Utils.ShowException(ex);
				}
			}
		}

		private void OnExtractionCompleted(IAsyncResult r)
		{
			if (InvokeRequired)
			{
				BeginInvoke(new AsyncCallback(OnExtractionCompleted), r);
			}
			else
			{
				NBiometricStatus status = NBiometricStatus.None;
				try
				{
					status = _biometricClient.EndCreateTemplate(r);
					if (status != NBiometricStatus.Ok)
					{
						MessageBox.Show(string.Format("Template was not extracted: {0}.", status), Text, MessageBoxButtons.OK, MessageBoxIcon.Error);
					}
				}
				catch (Exception ex)
				{
					Utils.ShowException(ex);
				}
				finally
				{
					if (status == NBiometricStatus.Ok)
					{
						btnIdentify.Enabled = _subjects != null;
					}
					else
					{
						_subject = null;
						btnIdentify.Enabled = false;
					}
				}
			}
		}

		private void SetFar()
		{
			try
			{
				_biometricClient.MatchingThreshold = Utils.MatchingThresholdFromString(cbMatchingFar.Text);
				cbMatchingFar.Text = Utils.MatchingThresholdToString(_biometricClient.MatchingThreshold);
			}
			catch
			{
				cbMatchingFar.Select();
				MessageBox.Show(@"FAR is not valid", Text, MessageBoxButtons.OK, MessageBoxIcon.Error);
			}
		}

		private void SetDefaultFar()
		{
			cbMatchingFar.Text = Utils.MatchingThresholdToString(_defaultFar);
			btnDefault.Enabled = false;
			SetFar();
		}

		private bool IsSubjectValid(NSubject subject)
		{
			return subject != null && (subject.Status == NBiometricStatus.Ok
				|| subject.Status == NBiometricStatus.None && subject.GetTemplateBuffer() != null);
		}

		#endregion

		#region Private form events

		private void IdentifyIrisLoad(object sender, EventArgs e)
		{
			lblFileForIdentification.Text = string.Empty;
			lblTemplatesCount.Text = @"0";

			try
			{
				cbMatchingFar.BeginUpdate();
				cbMatchingFar.Items.Add(0.001.ToString("P1"));
				cbMatchingFar.Items.Add(0.0001.ToString("P2"));
				cbMatchingFar.Items.Add(0.00001.ToString("P3"));
			}
			finally
			{
				cbMatchingFar.EndUpdate();
			}
		}

		private void BtnOpenTemplatesClick(object sender, EventArgs e)
		{
			btnIdentify.Enabled = false;
			_enrollStatus = NBiometricStatus.None;
			_subjects = null;

			lblTemplatesCount.Text = @"0";
			openFileDialog.Multiselect = true;
			openFileDialog.FileName = null;
			openFileDialog.Filter = string.Empty;
			openFileDialog.Title = @"Open Templates Files";
			if (openFileDialog.ShowDialog() == DialogResult.OK)
			{
				try
				{
					int templatesCount = openFileDialog.FileNames.Length;
					_biometricClient.Clear(); // Clear previously enrolled subjects
					_subjects = new NSubject[templatesCount];
					for (int i = 0; i < templatesCount; ++i)
					{
						_subjects[i] = NSubject.FromFile(openFileDialog.FileNames[i]);
						_subjects[i].Id = Path.GetFileName(openFileDialog.FileNames[i]);
					}
					lblTemplatesCount.Text = templatesCount.ToString(CultureInfo.InvariantCulture);

					NBiometricTask task = _biometricClient.CreateTask(NBiometricOperations.Enroll, null);
					foreach (NSubject item in _subjects)
					{
						task.Subjects.Add(item);
					}
					_biometricClient.BeginPerformTask(task, OnEnrollCompleted, null);
				}
				catch (Exception ex)
				{
					Utils.ShowException(ex);
				}
			}
		}

		private void BtnOpenImageClick(object sender, EventArgs e)
		{
			btnIdentify.Enabled = false;

			openFileDialog.FileName = null;
			openFileDialog.Filter = string.Empty;
			openFileDialog.Title = @"Open Image File Or Template";

			irisView.Iris = null;
			lblFileForIdentification.Text = string.Empty;
			_subject = null;
			if (openFileDialog.ShowDialog() == DialogResult.OK)
			{
				lblFileForIdentification.Text = openFileDialog.FileName;

				// Check if given file is a template
				try
				{
					_subject = NSubject.FromFile(openFileDialog.FileName);
					btnIdentify.Enabled = _subjects != null;
				}
				catch { }

				// If file is not a template, try to load it as image
				if (_subject == null)
				{
					// Create an iris object
					var iris = new NIris { FileName = openFileDialog.FileName };
					irisView.Iris = iris;
					// Add the iris to a subject
					_subject = new NSubject();
					_subject.Irises.Add(iris);

					// Extract the template from the subject
					_biometricClient.BeginCreateTemplate(_subject, OnExtractionCompleted, null);
				}
			}
		}

		private void BtnIdentifyClick(object sender, EventArgs e)
		{
			listView.Items.Clear();
			if (_subject != null && _subjects != null && _subjects.Length > 0)
			{
				_biometricClient.BeginIdentify(_subject, OnIdentifyDone, null);
			}
		}

		private void CbMatchingFarEnter(object sender, EventArgs e)
		{
			btnDefault.Enabled = true;
		}

		private void BtnDefaultClick(object sender, EventArgs e)
		{
			cbMatchingFar.SelectedIndex = 1;
			btnDefault.Enabled = false;
		}

		private void CbMatchingFarLeave(object sender, EventArgs e)
		{
			SetFar();
		}

		private void IdentifyIrisVisibleChanged(object sender, EventArgs e)
		{
			if (Visible && _biometricClient != null)
			{
				SetDefaultFar();
			}
		}

		#endregion
	}
}

File: java/Samples/Biometrics/VB/SimpleIrisesSampleVB/IdentifyIris.vb
Imports System
Imports System.Globalization
Imports System.IO
Imports System.Linq
Imports System.Windows.Forms
Imports Neurotec.Biometrics
Imports Neurotec.Biometrics.Client
Imports Neurotec.Images

Partial Public Class IdentifyIris
	Inherits UserControl
#Region "Public constructor"

	Public Sub New()
		InitializeComponent()
	End Sub

#End Region

#Region "Private fields"

	Private _biometricClient As NBiometricClient
	Private _subject As NSubject
	Private _subjects() As NSubject
	Private _enrollStatus As NBiometricStatus = NBiometricStatus.None

#End Region

#Region "Public properties"

	Public Property BiometricClient() As NBiometricClient
		Get
			Return _biometricClient
		End Get
		Set(ByVal value As NBiometricClient)
			_biometricClient = value
		End Set
	End Property

#End Region

#Region "Private methods"

	Private Sub OnEnrollCompleted(ByVal r As IAsyncResult)
		If InvokeRequired Then
			BeginInvoke(New AsyncCallback(AddressOf OnEnrollCompleted), r)
		Else
			Dim task As NBiometricTask = _biometricClient.EndPerformTask(r)
			_enrollStatus = task.Status
			If _enrollStatus = NBiometricStatus.Ok Then
				btnIdentify.Enabled = IsSubjectValid(_subject)
			Else
				MessageBox.Show(String.Format("Enrollment failed: {0}", _enrollStatus))
			End If
		End If
	End Sub

	Private Sub OnIdentifyDone(ByVal r As IAsyncResult)
		If InvokeRequired Then
			BeginInvoke(New AsyncCallback(AddressOf OnIdentifyDone), r)
		Else
			Try
				Dim status As NBiometricStatus = _biometricClient.EndIdentify(r)
				If status = NBiometricStatus.Ok OrElse status = NBiometricStatus.MatchNotFound Then
					' Matching subjects
					For Each result In _subject.MatchingResults
						listView.Items.Add(New ListViewItem(New String() {result.Id, result.Score.ToString(CultureInfo.InvariantCulture)}))
					Next result
					For Each subject In _subjects
						Dim tmpSubject As NSubject = subject
						Dim isMatchingResult As Boolean = _subject.MatchingResults.Any(Function(result) tmpSubject.Id = result.Id)
						If (Not isMatchingResult) Then
							listView.Items.Add(New ListViewItem(New String() {subject.Id, "0"}))
						End If
					Next subject
				Else
					MessageBox.Show(String.Format("Identification failed: {0}", status))
				End If
			Catch ex As Exception
				Utils.ShowException(ex)
			End Try
		End If
	End Sub

	Private Sub OnExtractionCompleted(ByVal r As IAsyncResult)
		If InvokeRequired Then
			BeginInvoke(New AsyncCallback(AddressOf OnExtractionCompleted), r)
		Else
			Dim status As NBiometricStatus = NBiometricStatus.None
			Try
				status = _biometricClient.EndCreateTemplate(r)
				If status <> NBiometricStatus.Ok Then
					MessageBox.Show(String.Format("Template was not extracted: {0}.", status), Text, MessageBoxButtons.OK, MessageBoxIcon.Error)
				End If
			Catch ex As Exception
				Utils.ShowException(ex)
			Finally
				If status = NBiometricStatus.Ok Then
					btnIdentify.Enabled = _subjects IsNot Nothing
				Else
					_subject = Nothing
					btnIdentify.Enabled = False
				End If
			End Try
		End If
	End Sub

	Private Sub SetFar()
		Try
			_biometricClient.MatchingThreshold = Utils.MatchingThresholdFromString(cbMatchingFar.Text)
			cbMatchingFar.Text = Utils.MatchingThresholdToString(_biometricClient.MatchingThreshold)
		Catch
			cbMatchingFar.Select()
			MessageBox.Show("FAR is not valid", Text, MessageBoxButtons.OK, MessageBoxIcon.Error)
		End Try
	End Sub

	Private Sub SetDefaultFar()
		cbMatchingFar.SelectedIndex = 1
		btnDefault.Enabled = False
		SetFar()
	End Sub

	Private Function IsSubjectValid(ByVal subject As NSubject) As Boolean
		Return subject IsNot Nothing AndAlso (subject.Status = NBiometricStatus.Ok OrElse subject.Status = NBiometricStatus.None AndAlso subject.GetTemplateBuffer() IsNot Nothing)
	End Function

#End Region

#Region "Private form events"

	Private Sub IdentifyIrisLoad(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
		lblFileForIdentification.Text = String.Empty
		lblTemplatesCount.Text = "0"

		Try
			cbMatchingFar.BeginUpdate()
			cbMatchingFar.Items.Add(0.001.ToString("P1"))
			cbMatchingFar.Items.Add(0.0001.ToString("P2"))
			cbMatchingFar.Items.Add(0.00001.ToString("P3"))
		Finally
			cbMatchingFar.EndUpdate()
		End Try
		cbMatchingFar.SelectedIndex = 1
	End Sub

	Private Sub BtnOpenTemplatesClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnOpenTemplates.Click
		btnIdentify.Enabled = False
		_enrollStatus = NBiometricStatus.None
		_subjects = Nothing

		lblTemplatesCount.Text = "0"
		OpenFileDialog.Multiselect = True
		OpenFileDialog.FileName = Nothing
		openFileDialog.Filter = Nothing
		OpenFileDialog.Title = "Open Templates Files"
		If OpenFileDialog.ShowDialog() = DialogResult.OK Then
			Try
				Dim templatesCount As Integer = OpenFileDialog.FileNames.Length
				_biometricClient.Clear() ' Clear previously enrolled subjects
				_subjects = New NSubject(templatesCount - 1) {}
				For i As Integer = 0 To templatesCount - 1
					_subjects(i) = NSubject.FromFile(OpenFileDialog.FileNames(i))
					_subjects(i).Id = Path.GetFileName(OpenFileDialog.FileNames(i))
				Next i
				lblTemplatesCount.Text = templatesCount.ToString(CultureInfo.InvariantCulture)

				Dim task As NBiometricTask = _biometricClient.CreateTask(NBiometricOperations.Enroll, Nothing)
				For Each item As NSubject In _subjects
					task.Subjects.Add(item)
				Next item
				_biometricClient.BeginPerformTask(task, AddressOf OnEnrollCompleted, Nothing)
			Catch ex As Exception
				Utils.ShowException(ex)
			End Try
		End If
	End Sub

	Private Sub BtnOpenImageClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnOpenImage.Click
		btnIdentify.Enabled = False

		OpenFileDialog.FileName = Nothing
		OpenFileDialog.Filter = String.Empty
		OpenFileDialog.Title = "Open Image File Or Template"

		irisView.Iris = Nothing
		lblFileForIdentification.Text = String.Empty
		_subject = Nothing
		If OpenFileDialog.ShowDialog() = DialogResult.OK Then
			lblFileForIdentification.Text = OpenFileDialog.FileName

			' Check if given file is a template
			Try
				_subject = NSubject.FromFile(OpenFileDialog.FileName)
				btnIdentify.Enabled = _subjects IsNot Nothing
			Catch
			End Try

			' If file is not a template, try to load it as image
			If _subject Is Nothing Then
				' Create an iris object
				Dim iris = New NIris With {.FileName = OpenFileDialog.FileName}
				irisView.Iris = iris
				' Add the iris to a subject
				_subject = New NSubject()
				_subject.Irises.Add(iris)

				' Extract the template from the subject
				_biometricClient.BeginCreateTemplate(_subject, AddressOf OnExtractionCompleted, Nothing)
			End If
		End If
	End Sub

	Private Sub BtnIdentifyClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnIdentify.Click
		ListView.Items.Clear()
		If _subject IsNot Nothing AndAlso _subjects IsNot Nothing AndAlso _subjects.Length > 0 Then
			_biometricClient.BeginIdentify(_subject, AddressOf OnIdentifyDone, Nothing)
		End If
	End Sub

	Private Sub CbMatchingFarEnter(ByVal sender As Object, ByVal e As EventArgs) Handles cbMatchingFar.Enter
		btnDefault.Enabled = True
	End Sub

	Private Sub BtnDefaultClick(ByVal sender As Object, ByVal e As EventArgs) Handles btnDefault.Click
		cbMatchingFar.SelectedIndex = 1
		btnDefault.Enabled = False
	End Sub

	Private Sub CbMatchingFarLeave(ByVal sender As Object, ByVal e As EventArgs) Handles cbMatchingFar.Leave
		SetFar()
	End Sub

	Private Sub IdentifyIrisVisibleChanged(ByVal sender As Object, ByVal e As EventArgs)
		If Visible AndAlso _biometricClient IsNot Nothing Then
			SetDefaultFar()
		End If
	End Sub

#End Region
End Class
